src/utils/polygon.cpp

```diff
@@ -370,21 +399,63 @@ void PolygonRef::simplify(const coord_t smallest_line_segment_squared, const coo
         //h^2 = L^2 / b^2     [factor the divisor]
         const coord_t height_2 = area_removed_so_far * area_removed_so_far / base_length_2;
         if ((height_2 <= 1 //Almost exactly colinear (barring rounding errors).
-            && LinearAlg2D::getDist2FromLine(current, previous, next) <= 1) // make sure that height_2 is not small because of cancellation of positive and negative areas
-            || (length2 < smallest_line_segment_squared
-                && next_length2 < smallest_line_segment_squared // Segments are small
-                && height_2 <= allowed_error_distance_squared) // removing the vertex doesn't introduce too much error.
-        )
+            && LinearAlg2D::getDist2FromLine(current, previous, next) <= 1)) // make sure that height_2 is not small because of cancellation of positive and negative areas
         {
-            continue; //Remove the vertex.
+            continue;
         }
+  
+        if (length2 < smallest_line_segment_squared
+            && height_2 <= allowed_error_distance_squared) // removing the vertex doesn't introduce too much error.) // removing the vertex doesn't introduce too much error.
+        {
+            if(next_length2 > smallest_line_segment_squared)
+            {
+                // Special case; The next line is long. If we were to remove this, it could happen that we get quite noticeable artifacts.
+                // We should instead move this point to a location where both edges are kept and then remove the previous point that we wanted to keep.
+                // By taking the intersection of these two lines, we get a point that perseves the direction (so it makes the corner a bit more pointy)
+                Point intersection_point;
+                bool has_intersection = lineLineIntersection(previous_previous, previous, current, next, &intersection_point);
 
-        //Don't remove the vertex.
+                if(has_intersection)
+                {
+                    // Find out if it is a degenerate intersection.
+                    if(vSize2(current - intersection_point) > allowed_error_distance_squared)
+                    {
+                        // The intersection is way to far away. Drop it. 
+                        continue;
```

Comment:
I'm not sure this should be a `continue`. Note that this code is in the case where the next segment is too long to be able to remove the vertex according to criterium 1. 

![image](https://user-images.githubusercontent.com/8895761/92921672-c9d6dd80-f434-11ea-8d71-e321b2f420b3.png)
black = original; red = simplified according to this line.

Reply:
The distance is only too long if the lines are almost paralel. At that point, it will create an intersection that is very far away, which results in weird ass spikes being put there.

Reply:
There are 3 alternatives: either we A) remove the current vertex, or B) we remove the current and last vertex and use the intersection point, or we C) leave both in. In this if-case you opt for A, but I am saying we should opt for C. You misinterpreted my comment to mean that we should opt for B.

Reply:
Possible 4th alternative: cut a long segment so that the shortcut is as long as the allowed error distance (or smallest segment length ?)
![image](https://user-images.githubusercontent.com/8895761/92933489-98670d80-f446-11ea-84ca-da7dc81ba940.png)

Reply:
Current solution (A):
![CurrentSolution](https://user-images.githubusercontent.com/3922611/93208177-74643e80-f75c-11ea-8d82-9a9bcecba338.png)
Leave both in (C):
![KeepBoth](https://user-images.githubusercontent.com/3922611/93208233-91990d00-f75c-11ea-88a9-c475caec54b3.png)

Option C is retaining some of the weird points. I also think that my previous assessment is incorrect; It happens to be wrong if both of the lines are almost in the same direction; It's probably because of a rounding error that the found point is very far away.

Reply:
As discussed with @Ghostkeeper, we've opted to go for option E.

It's basically option C, but when the line segment is below 5 mu, we still discard it.

Reply:
This whole function was built to remove line segments below 5mu, but to do it in a safe way. If you choose option E then we lose the safety this function should guarantee. We might introduce escalation problems again where a lot of consecutive 5mu line segments get removed and introduce a too big error.